p2p/transport/webrtc/udpmux/mux.go

```diff
@@ -0,0 +1,308 @@
+// remote address. If found we forward the packet to the connection. If an associated
+// connection is not found, we check to see if the packet is a STUN packet. We then
+// fetch the ufrag of the remote from the STUN packet and use it to check if there
+// is a connection associated with the (ufrag, IP address family) pair. If found
+// we add the association to the address map. If not found, it is a previously
+// unseen IP address and the `unknownUfragCallback` callback is invoked.
+type udpMux struct {
+	socket               net.PacketConn
+	unknownUfragCallback func(string, net.Addr)
+
+	storage *udpMuxStorage
+
+	// the context controls the lifecycle of the mux
+	wg     sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func NewUDPMux(socket net.PacketConn, unknownUfragCallback func(string, net.Addr)) ice.UDPMux {
+	ctx, cancel := context.WithCancel(context.Background())
+	mux := &udpMux{
+		ctx:                  ctx,
+		cancel:               cancel,
+		socket:               socket,
+		unknownUfragCallback: unknownUfragCallback,
+		storage:              newUdpMuxStorage(),
+	}
+
+	mux.wg.Add(1)
+	go func() {
+		defer mux.wg.Done()
+		mux.readLoop()
+	}()
+	return mux
+}
+
+// GetListenAddresses implements ice.UDPMux
+func (mux *udpMux) GetListenAddresses() []net.Addr {
+	return []net.Addr{mux.socket.LocalAddr()}
+}
+
+// GetConn implements ice.UDPMux
+// It creates a net.PacketConn for a given ufrag if an existing
+// one cannot be  found. We differentiate IPv4 and IPv6 addresses
+// as a remote is capable of being reachable through multiple different
+// UDP addresses of the same IP address family (eg. Server-reflexive addresses
+// and peer-reflexive addresses).
+func (mux *udpMux) GetConn(ufrag string, addr net.Addr) (net.PacketConn, error) {
+	a, ok := addr.(*net.UDPAddr)
+	if !ok && addr != nil {
+		return nil, fmt.Errorf("unexpected address type: %T", addr)
+	}
+	isIPv6 := ok && a.IP.To4() == nil
+	return mux.getOrCreateConn(ufrag, isIPv6)
+}
+
+// Close implements ice.UDPMux
+func (mux *udpMux) Close() error {
+	select {
+	case <-mux.ctx.Done():
+		return nil
+	default:
+	}
+	mux.cancel()
+	mux.socket.Close()
+	mux.wg.Wait()
+	return nil
+}
+
+// RemoveConnByUfrag implements ice.UDPMux
+func (mux *udpMux) RemoveConnByUfrag(ufrag string) {
+	if ufrag != "" {
+		mux.storage.RemoveConnByUfrag(ufrag)
+	}
+}
+
+func (mux *udpMux) getOrCreateConn(ufrag string, isIPv6 bool) (net.PacketConn, error) {
+	select {
+	case <-mux.ctx.Done():
+		return nil, io.ErrClosedPipe
+	default:
+		return mux.storage.GetOrCreateConn(ufrag, isIPv6, mux)
+	}
+}
+
+// writeTo writes a packet to the underlying net.PacketConn
+func (mux *udpMux) writeTo(buf []byte, addr net.Addr) (int, error) {
+	return mux.socket.WriteTo(buf, addr)
+}
+
+func (mux *udpMux) readLoop() {
+	for {
+		select {
+		case <-mux.ctx.Done():
+			return
+		default:
+		}
+
+		buf := pool.Get(ReceiveMTU + 2048)
```

Comment:
Quick TODO: I added 2048 to the receive MTU here to avoid having both this loop and the noise library fetch from the same global pool. This was causing a race condition as documented [here](https://github.com/libp2p/go-libp2p/pull/1655#issuecomment-1374818177). This fetches a much larger buffer than required, so it should be okay to just use `ReceiveMTU`.

Reply:
Ok nice. I've removed it.